blog: write more bash (#909)
tilt-dev/tilt.build: blog/_posts/2021-08-16-write-more-bash.md

---
slug: "write-more-bash"
date: 2021-08-16
author: nick
layout: blog
title: "Write More Bash to Hack Features Faster and with Less Testing"
subtitle: "A walkthrough of the 'cancel' button extension"
description: "A walkthrough of the 'cancel' button extension"
image: "/assets/images/write-more-bash/bonk.jpg"
image_type: "contain"
tags:
  - api
  - bash
---

The KubeCon schedule was announced last week. The Tilt team is giving two talks
😇.

- ["Beyond Kubernetes Security"](https://sched.co/lV4f) - A "blockbuster action thriller."
- ["The Control Loop as an Application Development Framework"](https://sched.co/lV1E) - How to write apps with more control loops.

Both talks are tangentially related to Tilt at best! They're mostly about
Kubernetes, the ideas behind Kubernetes, and how those ideas can (and should!)
affect your life as a dev.

In the Control Loop talk, I'm going to talk more about the Kubernetes ideas and
libraries that Tilt is built on top of. The Kubernetes libraries have a lot of

> tools to help us build robust systems.  But even better, they also have a lot of

Reviewer: Assuming the "but even better" is a joke, I missed it on the first read. If you wanna get less subtle (which is obviously always a balance with humor), you could do something like this:
```suggestion
tools to help us build robust systems, which, sure, some people find value in.  But far, far more importantly, they also have a lot of
```

Author: :stuck_out_tongue:

tools that let you hack together shit fast in Bash.

What does this look like when we're developing Tilt? I'm going to show you a
Bash-based Tilt extension I wrote last week! And as we go through it, you'll learn:

- Why Kubernetes stuff lends itself so well to Bash
- What a Bash-based Tilt extension looks like
- How you too can hack stuff into Tilt with Bash

## Why Kubernetes stuff lends itself so well to Bash

The core idea of Kubernetes is simple: if you want to manage servers, you need a way
to react to changes at runtime.

On top of that, Kubernetes offers a suite of data models, HTTP APIs, and CLIs
that are really consistent and well thought-out! (See: ["Kubernetes Is So Simple
You Can Explore it with
Curl"](https://blog.tilt.dev/2021/03/18/kubernetes-is-so-simple.html).)

This API platform allows teams to create very complex, robust features.
But few people appreciate that you can also use it to build simple, hacky features!

A few weeks ago [L](https://twitter.com/ellenkorbes) asked if we could add a
cancel button to the Tilt UI to kill a running local server. And I realized that you could

> hack this together with a couple dozen lines of Bash without modifying core Tilt! Here it is!

Reviewer: it seems worth adding a transition from "the k8s api is cool" to "I did a thing in Tilt" along the lines of "Tilt implements its API using the k8s libraries"

Author: done

> hack this together with a couple dozen lines of Bash without modifying core Tilt! Here it is!

Reviewer: Something like this might be appealing, but feel free to ignore:
"By exposing its internals via an API, Tilt makes easy to add/prototype features - you don't have to worry about getting code into main before you can use it, or breaking existing code, and you can write in any language you want. In fact, you can even use bash!"

Author: done

the implementation code behind a feature.

Next, we write [a basic
loop](https://github.com/tilt-dev/tilt-extensions/blob/master/cancel/cancel_btn_controller.sh)
in Bash that watches objects for changes:

```bash
#!/bin/bash
#
# Starts a controller that watches for new Cmds, and
# creates buttons to cancel them.

set -eou pipefail

echo "operator:cancel runs in the background and listens to Tilt
When there are commands to cancel, operator:cancel adds a Cancel button to the Tilt UI
"

tilt get cmd --watch -o name | while read -r cmd_full_name; do
    cmd_short_name=${cmd_full_name#cmd.tilt.dev/}
    ./reconcile_cancel_btn.sh "$cmd_short_name"
done
```

This code uses Tilt's [Cmd API](https://api.tilt.dev/core/cmd-v1alpha1.html) to
watch all local commands. Tilt's CLI reuses large parts of `kubectl`, and has
many of the same verbs.  You can use `tilt api-resources` to look at all the
available objects, `tilt get` to view a summary of those objects, and `tilt get
--watch` to watch for updates to those objects.

This Bash loop streams Cmd changes and passes them to a "reconciler."

"Reconciler" is another common phrase in the Kubernetes ecosystem for the part

> of a controller that compares state to the desired state. You can read more about reconcilers

Reviewer: 1. tie it back to the three steps listed above
2. it also does step (3), yeah?

```suggestion
of a controller that implements steps (2) and (3) of the reactive control loop above: it compares the live state to the desired state, and applies changes to the live state to bring it to the desired state. You can read more about reconcilers
```

Author: done

handles creating the button but not deleting the button. (You can read [the full
source
code](https://github.com/tilt-dev/tilt-extensions/blob/master/cancel/reconcile_cancel_btn.sh)
to see how we delete buttons.)

```bash
cancel_cmd_name="$cmd_name:cancel"
cancel_button_name="$cmd_name:cancel"
cmd=$(tilt get cmd "$cmd_name" -o json --ignore-not-found)

# If the command isn't running or doesn't exist, disable the button
pid=$(echo "$cmd" | jq -r '.status.running.pid')
disabled="false"
if [[ "$pid" == "" || "$pid" == "null" ]]; then
    disabled="true"
fi

dir=$(realpath "$(dirname "$0")")
cat <<EOF | tilt apply -f -
apiVersion: tilt.dev/v1alpha1
kind: UIButton
metadata:
  name: $cancel_button_name
spec:
  disabled: $disabled
  text: Cancel
  location:
    componentType: resource
    componentID: $resource
---
apiVersion: tilt.dev/v1alpha1
kind: Cmd
metadata:
  name: $cancel_cmd_name
  annotations:
    "tilt.dev/resource": "$resource"
    "tilt.dev/log-span-id": "$cancel_cmd_name"
spec:
  args: ["./kill_cmd.sh", "$cmd_name"]
  dir: $dir
  startOn:
    uiButtons:
    - $cancel_button_name
EOF
```

The `cancel` command and the button that triggers this `cancel` command are

> objects. They're specified like another other object in the Kubernetes

Reviewer 2: ```suggestion
objects. They're specified like any other object in the Kubernetes
```

ecosystem - as YAML. Because our control loop reacts to all changes, it can

> disable the `cancel` button when the command exists.

Reviewer 2: Did you mean s/exists/exits? 🤔
